File: 10.BigFibonacci/code/fibonacci.py
rfile = open("fibonacci.inp", "r")
wfile = open("fibonacci.out", "w")

def get_pisano_period(M) :
    M = int(M)

    a = 0
    b = 1
    c = a + b

    for i in range(M * M) :
        c = (a + b) % M
        a = b
        b = c

        if a == 0 and b == 1 : return i + 1


# def get_fibonacci_formula(n, m, pisano):
#     n = int(n)
#     m = int(m)

#     n = n % pisano

#     print(n, m, pisano)
#     sqrt_5 = 5 ** (1/2)
#     ans = 1 / sqrt_5 * ( ((1 + sqrt_5) / 2) ** n  - ((1 - sqrt_5) / 2) ** n )
#     return int(ans) % m

# def get_fibonacci_huge(n, m, pisano) :

#     n = int(n)
#     m = int(m)
#     pisano = int(pisano)

#     if n < 2 : return n
#     remainder = n % pisano

#     first = 0
#     second = 1

#     res = remainder

#     for i in range(remainder - 1) :
#         res = (first + second) % m
#         first = second % m
#         second = res % m

#     return res


def get_fibonacci_pisano(m, pisano) :

    m = int(m)
    pisano = int(pisano)
    ret = {}
    ret[0] = 0
    ret[1] = 1

    first = 0
    second = 1

    for i in range(pisano - 1) :
        res = (first + second) % m
        first = second % m
        second = res
        ret[i+2] = res

    return ret

# pisano_period = get_fibonacci_pisano(3, 8)
# print(pisano_period)

# n = 10000000
# print(pisano_period[n%3])
# import sys
# sys.exit(0)


while True :
    line = str(rfile.readline()).strip()

    if line == "" : break

    for i in range(int(line)) :
        print("-------------")
        line = str(rfile.readline()).strip()

        (fA, fB, fC, fD, M) = line.split(" ")
        # print(fA, " ", fB, " ", fC, " ", fD, " ", M)

        pisano = get_pisano_period(M)
        # print("피사노 주기 : ", pisano)

        resultA = get_fibonacci_huge(fA, M, pisano)
        resultB = get_fibonacci_huge(fB, M, pisano)
        resultC = get_fibonacci_huge(fC, M, pisano)
        resultD = get_fibonacci_huge(fD, M, pisano)

        print(resultA, " ", resultB, " ", resultC, " ", resultD, " ", M)

        M = int(M)
        print(((resultA * resultB) + (resultC * resultD)) % M)
        # ((fa % m * fb % m ) % m + (fc %m * fd % m) % m ) % m
        print("result : ", ( ((resultA * resultB) % M) + ((resultC * resultD) % M)) % M )
        # result = (((resultA * resultB) % M) + ((resultC * resultD) % M)) % M
        # print("result : ", result)
        # wfile.write("%s\n" % result)

rfile.close()
wfile.close()

File: 10.BigFibonacci/test/test/fibonacci.cpp
#pragma warning(disable: 4996)
#include <iostream>

using namespace std;

/*
item = [0,0,0,0]

list 
 length = 3
 capacity = 8
 items = [0, 1, 2, 3, 4, null, null, null]

add(list, 0)
add(list ,1)
add(list, 2)
add(list, 3)
add(list, 4)
*/

struct list {
    int length;
    unsigned long long capacity;
    int* items;
};

typedef struct list list;

// Make a new list
list* newList(unsigned long long pisano) {
    list* al = (list * )malloc(sizeof(list));
    int* items = (int *)malloc(pisano * sizeof(int));
    *al = (list) { 0, pisano, items };
    return al;
}

// Expand list
//void expand(list* al) {
//    al->capacity = al->capacity * 3 / 2;
//    al->items = (int *)realloc(al->items, al->capacity * sizeof(int));
//}

// Add an int to list
void add(list* al, int n) {
    //if (al->length >= al->capacity) expand(al);
    al->items[al->length] = n;
    al->length++;
}

// Print list
void print(list* al) {
    printf("[ ");
    for (int i = 0; i < al->length; i++) {
        if (i > 0) printf(", ");
        printf("%d", al->items[i]);
    }
    printf(" ]\n");
}

unsigned long long get_pisano_period(int M) {
    unsigned long long a = 0;
    unsigned long long b = 1;
    unsigned long long c = a + b;

    for (unsigned long long i = 0; i < M * M; i++) {
        c = (a + b) % M;
        a = b;
        b = c;

        if (a == 0 && b == 1) {
            return i + 1;
        }
    }
}

void get_fibonacci_pisano(int m, int pisano, list* pisanoList) {

    add(pisanoList, 0);
    add(pisanoList, 1);

    int first = 0;
    int second = 1;

    //printf("pisano : %d\n", pisano);
    for (int i = 0; i < pisano - 1; i++) {
        int res = (first + second) % m;
        first = second % m;
        second = res;
        add(pisanoList, res);
    }
}

//int get_fibonacci_huge(int n, int m, int pisano) {
//
//    if (n < 2) { return n; }
//
//    int remainder = n % pisano;
//
//    int first = 0;
//    int second = 1;
//
//    int res = remainder;
//
//    for (int i = 0; i < remainder - 1; i++) {
//        res = (first + second) % m;
//        first = second % m;
//        second = res % m;
//    }
//
//    return res;
//}

int main() {
    FILE* inp;
    FILE* out;

    inp = fopen("fibonacci.inp", "rt");
    out = fopen("fibonacci.out", "wt");

    int numOfCase = 0;
    if (fscanf(inp, "%d", &numOfCase) == 1) {}

    list* pisanoList;
    

    //printf("%d", numOfCase);
    for (int i = 0; i < numOfCase; i++) {
        //printf("\n---------------\n");

        int fA = 0, fB = 0, fC = 0, fD = 0, M = 0;

        if (fscanf(inp, "%d %d %d %d %d", &fA, &fB, &fC, &fD, &M) == 1) {}
        //printf("%d %d %d %d %d\n", fA, fB, fC, fD, M);

        unsigned long long pisano = get_pisano_period(M);
        //printf("_pisano : %llu\n", pisano);
        pisanoList = newList(pisano);

        /*unsigned long long resultA = (unsigned long long)get_fibonacci_huge(fA, M, pisano);
        unsigned long long resultB = (unsigned long long)get_fibonacci_huge(fB, M, pisano);
        unsigned long long resultC = (unsigned long long)get_fibonacci_huge(fC, M, pisano);
        unsigned long long resultD = (unsigned long long)get_fibonacci_huge(fD, M, pisano);

        printf("result A, B, C, D =>  %llu %llu %llu %llu\n", resultA, resultB, resultC, resultD);*/

        //unsigned long long result = (((resultA * resultB) % M) + ((resultC * resultD) % M)) % M;
        //result = (((resultA * resultB)) % M + ((resultC * resultD)) % M) % M

        //printf("result : %llu\n", result);
        //fprintf(out, "%llu\n", result);

        get_fibonacci_pisano(M, pisano, pisanoList);
        //print(pisanoList);

        unsigned long long resultA = (unsigned long long)(pisanoList->items[fA % pisano]);
        unsigned long long resultB = (unsigned long long)(pisanoList->items[fB % pisano]);
        unsigned long long resultC = (unsigned long long)(pisanoList->items[fC % pisano]);
        unsigned long long resultD = (unsigned long long)(pisanoList->items[fD % pisano]);

        //printf("result A, B, C, D =>  %llu %llu %llu %llu\n", resultA, resultB, resultC, resultD);

        unsigned long long result = (((resultA * resultB)) % M + ((resultC * resultD)) % M) % M;
        //printf("result : %llu\n", result);

        cout >> result >> endl;

        fprintf(out, "%llu\n", result);
    }

    fclose(inp);
    fclose(out);

    return 0;
}